src/rust/engine/src/externs/options.rs

```diff
@@ -164,11 +165,8 @@ struct PyOptionParser(OptionParser);
 // The pythonic value of a dict-typed option.
 type PyDictVal = HashMap<String, PyObject>;
 
-// The derivation of the option value, as a vec of (value, vec of source ranks) tuples.
-// A scalar value will always have a single source. A list/dict value may have elements
-// appended across multiple sources.
-// In ascending rank order, so the last value is the final value of the option.
-type OptionValueDerivation<T> = Vec<(T, Vec<isize>)>;
+// The derivation of the option value, as a vec of (value, rank, details string) tuples.
+type OptionValueDerivation<'py, T> = Vec<(T, isize, Option<Bound<'py, PyString>>)>;
```

Comment:
Note that we may not need that isize after we get rid of the legacy parser. The RankedValue struct on the Python side expects it, but I'm not sure we actually use it.